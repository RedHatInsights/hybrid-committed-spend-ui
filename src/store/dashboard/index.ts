import type { DashboardWidget } from './dashboardCommon';
import { DashboardSize, dashboardStateKey } from './dashboardCommon';
import { dashboardReducer } from './dashboardReducer';
import * as dashboardSelectors from './dashboardSelectors';

export { dashboardStateKey, dashboardReducer, dashboardSelectors, DashboardSize };
export type { DashboardWidget };
