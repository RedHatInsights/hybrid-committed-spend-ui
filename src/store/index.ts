export { RootState } from './rootReducer';
export { configureStore } from './store';
