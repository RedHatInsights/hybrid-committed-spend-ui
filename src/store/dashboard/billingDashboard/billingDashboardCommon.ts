import { BillingFilters, BillingQuery, getBillingQuery } from 'api/queries';
import { DashboardWidget } from 'store/dashboard/common';

export const billingDashboardStateKey = 'billingDashboard';
export const billingDashboardDefaultFilters: BillingFilters = {
  time_scope_units: 'month',
  time_scope_value: -1,
  resolution: 'daily',
};

// eslint-disable-next-line no-shadow
export const enum BillingCardType {
  actualSpend = 'actualSpend',
  actualSpendBreakdown = 'actualSpendBreakdown',
  committedSpend = 'committedSpend',
  committedSpendTrend = 'committedSpendTrend',
}

export interface BillingDashboardWidget extends DashboardWidget<BillingCardType> {}

export function getQueryForWidget(filter: BillingFilters = billingDashboardDefaultFilters, props?) {
  const query: BillingQuery = {
    filter,
    ...(props ? props : {}),
  };
  return getBillingQuery(query);
}

export function getQueryForWidgetTabs(
  widget: BillingDashboardWidget,
  filter: BillingFilters = billingDashboardDefaultFilters
) {
  const query: BillingQuery = {
    filter,
    // group_by: getGroupByForTab(widget.currentTab),
  };
  return getBillingQuery(query);
}
