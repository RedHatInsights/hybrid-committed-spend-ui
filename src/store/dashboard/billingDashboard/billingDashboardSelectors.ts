import { RootState } from 'store/rootReducer';

import {
  billingDashboardDefaultFilters,
  billingDashboardStateKey,
  getQueryForWidget,
  getQueryForWidgetTabs,
} from './billingDashboardCommon';

export const selectBillingDashboardState = (state: RootState) => state[billingDashboardStateKey];

export const selectWidgets = (state: RootState) => selectBillingDashboardState(state).widgets;

export const selectWidget = (state: RootState, id: number) => selectWidgets(state)[id];

export const selectCurrentWidgets = (state: RootState) => selectBillingDashboardState(state).currentWidgets;

export const selectWidgetQueries = (state: RootState, id: number) => {
  const widget = selectWidget(state, id);

  const defaultFilter = {
    ...billingDashboardDefaultFilters,
    ...(widget.filter ? widget.filter : {}),
  };

  return {
    previous: getQueryForWidget({
      ...defaultFilter,
      time_scope_value: -2,
    }),
    current: getQueryForWidget(defaultFilter),
    forecast: getQueryForWidget({}, { limit: 31 }),
    tabs: getQueryForWidgetTabs(widget, {
      resolution: 'monthly',
    }),
  };
};
