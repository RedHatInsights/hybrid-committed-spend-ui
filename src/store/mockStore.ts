import type { DeepPartial, Reducer, ReducersMapObject, Store } from 'redux';
import { applyMiddleware, combineReducers, createStore } from 'redux';

import type { RootState } from './rootReducer';
import { middlewares } from './store';

type MockReducer<S> = ReducersMapObject<S, any> | Reducer<S>;

export function createMockStoreCreator<S>(reducer: MockReducer<S>) {
  const rootReducer = typeof reducer === 'object' ? combineReducers(reducer) : reducer;

  return (initialState?: DeepPartial<S>): Store<RootState, any> => {
    return createStore<RootState, any, any, any>(rootReducer as any, initialState, applyMiddleware(...middlewares));
  };
}
