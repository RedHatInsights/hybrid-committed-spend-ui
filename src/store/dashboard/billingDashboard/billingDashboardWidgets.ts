import { ReportPathsType, ReportType } from 'api/reports/report';
import messages from 'locales/messages';

import { BillingCardType, BillingDashboardWidget } from './billingDashboardCommon';

let currrentId = 0;
const getId = () => currrentId++;

export const actualSpendWidget: BillingDashboardWidget = {
  id: getId(),
  title: messages.dashboardActualSpendTitle,
  reportPathsType: ReportPathsType.billing,
  reportType: ReportType.cost,
  type: BillingCardType.actualSpend,
};

export const committedSpendWidget: BillingDashboardWidget = {
  id: getId(),
  title: messages.dashboardCommitmentBalanceTitle,
  reportPathsType: ReportPathsType.billing,
  reportType: ReportType.cost,
  type: BillingCardType.committedSpend,
};
