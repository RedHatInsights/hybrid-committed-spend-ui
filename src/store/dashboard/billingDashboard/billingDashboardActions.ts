import { ThunkAction } from 'store/common';
import { reportActions } from 'store/reports';
import { createAction } from 'typesafe-actions';

import { selectWidget, selectWidgetQueries } from './billingDashboardSelectors';

// export const fetchWidgetForecasts = (id: number): ThunkAction => {
//   return (dispatch, getState) => {
//     const state = getState();
//     const widget = selectWidget(state, id);
//
//     if (widget.forecastPathsType && widget.forecastType) {
//       const { forecast } = selectWidgetQueries(state, id);
//       dispatch(forecastActions.fetchForecast(widget.forecastPathsType, widget.forecastType, forecast));
//     }
//   };
// };

export const fetchWidgetReports = (id: number): ThunkAction => {
  return (dispatch, getState) => {
    const state = getState();
    const widget = selectWidget(state, id);
    const { /* previous, */ current } = selectWidgetQueries(state, id);
    dispatch(reportActions.fetchReport(widget.reportPathsType, widget.reportType, current));
    // dispatch(reportActions.fetchReport(widget.reportPathsType, widget.reportType, previous));
  };
};

export const setWidgetTab = createAction('ocpDashboard/widget/tab')<{
  id: number;
}>();
