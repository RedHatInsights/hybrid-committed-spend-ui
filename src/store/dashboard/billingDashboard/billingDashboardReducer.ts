import { ActionType, getType } from 'typesafe-actions';

import { setWidgetTab } from './billingDashboardActions';
import { BillingDashboardWidget } from './billingDashboardCommon';
import { actualSpendWidget, committedSpendWidget } from './billingDashboardWidgets';

export type BillingDashboardAction = ActionType<typeof setWidgetTab>;

export type BillingDashboardState = Readonly<{
  widgets: Record<number, BillingDashboardWidget>;
  currentWidgets: number[];
}>;

export const defaultState: BillingDashboardState = {
  currentWidgets: [actualSpendWidget.id, committedSpendWidget.id],
  widgets: {
    [actualSpendWidget.id]: actualSpendWidget,
    [committedSpendWidget.id]: committedSpendWidget,
  },
};

export function billingDashboardReducer(state = defaultState, action: BillingDashboardAction): BillingDashboardState {
  switch (action.type) {
    case getType(setWidgetTab):
      return {
        ...state,
        widgets: {
          ...state.widgets,
          [action.payload.id]: {
            ...state.widgets[action.payload.id],
          },
        },
      };
    default:
      return state;
  }
}
