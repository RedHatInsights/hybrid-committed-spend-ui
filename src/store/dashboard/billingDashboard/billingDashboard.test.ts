import { actualSpendWidget, committedSpendWidget } from './billingDashboardWidgets';

jest.mock('store/reports/reportActions');

import { createMockStoreCreator } from 'store/mockStore';
import { reportActions } from 'store/reports';

import * as actions from './billingDashboardActions';
import { billingDashboardStateKey } from './billingDashboardCommon';
import { billingDashboardReducer } from './billingDashboardReducer';
import * as selectors from './billingDashboardSelectors';

const createOcpDashboardStore = createMockStoreCreator({
  [billingDashboardStateKey]: billingDashboardReducer,
});

const fetchReportMock = reportActions.fetchReport as jest.Mock;

beforeEach(() => {
  fetchReportMock.mockReturnValue({ type: '@@test' });
});

test('default state', () => {
  const store = createOcpDashboardStore();
  const state = store.getState();
  expect(selectors.selectCurrentWidgets(state)).toEqual([actualSpendWidget.id, committedSpendWidget.id]);
  expect(selectors.selectWidget(state, actualSpendWidget.id)).toEqual(actualSpendWidget);
});

test('fetch widget reports', () => {
  const store = createOcpDashboardStore();
  store.dispatch(actions.fetchWidgetReports(actualSpendWidget.id));
  expect(fetchReportMock.mock.calls).toMatchSnapshot();
});
