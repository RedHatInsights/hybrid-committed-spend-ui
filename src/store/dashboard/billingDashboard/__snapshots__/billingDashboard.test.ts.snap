// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch widget reports 1`] = `
[
  [
    "billing",
    "cost",
    "filter[time_scope_units]=month&filter[time_scope_value]=-1&filter[resolution]=daily",
  ],
]
`;
